3rdparty/hal_rvv/hal_rvv_1p0/dxt.hpp

```diff
@@ -0,0 +1,609 @@
+    };
+}
+
+// the algorithm is copied from core/src/dxt.cpp,
+// in the function template static void cv::DFT and cv::DFT_R2, cv::DFT_R3, cv::DFT_R5
+template<typename T>
+inline int dft(const Complex<T>* src, Complex<T>* dst, int nf, int *factors, T scale, int* itab,
+                  const Complex<T>* wave, int tab_size, int len, bool isInverse, bool noPermute)
+{
+    int n = len;
+    int f_idx, nx;
+    int dw0 = tab_size, dw;
+    int i, j, k;
+    Complex<T> t;
+
+    int tab_step = tab_size == n ? 1 : tab_size == n*2 ? 2 : tab_size/n;
+    int vl;
+
+    // 0. shuffle data
+    if( dst != src )
+    {
+        if( !isInverse )
+        {
+            for( i = 0; i < n; i += vl )
+            {
+                vl = dxt::rvv<T>::vsetvl_itab(n - i);
+                auto vec_itab = dxt::rvv<T>::vlse_itab(reinterpret_cast<const uint*>(itab + i * tab_step), sizeof(int) * tab_step, vl);
+                vec_itab = __riscv_vmul(vec_itab, sizeof(T) * 2, vl);
+                auto vec_src_re = __riscv_vloxei32(reinterpret_cast<const T*>(src), vec_itab, vl);
+                vec_itab = __riscv_vadd(vec_itab, sizeof(T), vl);
+                auto vec_src_im = __riscv_vloxei32(reinterpret_cast<const T*>(src), vec_itab, vl);
+                dxt::rvv<T>::vsse_itab(reinterpret_cast<T*>(dst + i), sizeof(T) * 2, vec_src_re, vl);
+                dxt::rvv<T>::vsse_itab(reinterpret_cast<T*>(dst + i) + 1, sizeof(T) * 2, vec_src_im, vl);
+            }
+        }
+        else
+        {
+            for( i = 0; i < n; i += vl )
+            {
+                vl = dxt::rvv<T>::vsetvl_itab(n - i);
+                auto vec_itab = dxt::rvv<T>::vlse_itab(reinterpret_cast<const uint*>(itab + i * tab_step), sizeof(int) * tab_step, vl);
+                vec_itab = __riscv_vmul(vec_itab, sizeof(T) * 2, vl);
+                auto vec_src_re = __riscv_vloxei32(reinterpret_cast<const T*>(src), vec_itab, vl);
+                vec_itab = __riscv_vadd(vec_itab, sizeof(T), vl);
+                auto vec_src_im = __riscv_vloxei32(reinterpret_cast<const T*>(src), vec_itab, vl);
+                vec_src_im = __riscv_vfneg(vec_src_im, vl);
+                dxt::rvv<T>::vsse_itab(reinterpret_cast<T*>(dst + i), sizeof(T) * 2, vec_src_re, vl);
+                dxt::rvv<T>::vsse_itab(reinterpret_cast<T*>(dst + i) + 1, sizeof(T) * 2, vec_src_im, vl);
+            }
+        }
+    }
+    else
+    {
+        // copied from core/src/dxt.cpp, it is slow to swap elements by intrinsics
+        if( !noPermute )
+        {
+            if( nf == 1 )
+            {
+                if( (n & 3) == 0 )
+                {
+                    int n2 = n/2;
+                    Complex<T>* dsth = dst + n2;
+
+                    for( i = 0; i < n2; i += 2, itab += tab_step*2 )
+                    {
+                        j = itab[0];
+
+                        CV_SWAP(dst[i+1], dsth[j], t);
```

Comment:
Technically this macro is not available for HAL implementations - only things from `*/hal/interface.h` should be used. It works because this HAL is header-only and is included after cvdef.h.